### DISCUSSION TITLE
[17일차 - 지원] 로그인 기능 jwt vs session

### DISCUSSION THREAD

**Discussion by smarfy99** [zzangCS/zzangInterview#55] [프론트 면접 질문]
jwt와 session 각각을 사용했을 때 장단점을 말해주세요.

**Comment by ez1n** [comment]
**1. 서버 상태 유무**
- Session 기반 인증: 서버는 세션 상태를 유지하고 클라이언트에게 세션 ID를 제공합니다. 세션 ID는 보통 쿠키를 통해 클라이언트에 저장됩니다. 서버는 세션 ID를 사용하여 클라이언트를 식별하고 해당 세션에 저장된 상태를 관리합니다.
- JWT: JWT는 자체적으로 상태를 갖지 않으며, 클라이언트 측에 정보를 저장합니다. 서버는 JWT의 서명을 통해 유효성을 검증할 뿐입니다. 따라서 서버에 세션 상태를 유지할 필요가 없으며, 확장성과 분산 시스템에서 더 유리한 점을 가지고 있습니다.


**2. 확장성과 분산 시스템**
- Session 기반 인증: 세션 기반 인증은 서버 상태를 유지해야 하므로 여러 서버 간의 세션 공유를 위한 추가적인 구성이 필요합니다. 이는 확장성과 분산 시스템에서 복잡성을 초래할 수 있습니다.
- JWT: JWT는 클라이언트에 정보를 저장하고 서버는 유효성 검증만 수행하므로, 서버 간의 세션 공유가 필요하지 않습니다. 따라서 분산 시스템에서 보다 용이하게 확장할 수 있습니다.


**3. 인증 서버의 부담**
- Session 기반 인증: 서버는 각 클라이언트의 세션 상태를 유지하고 관리해야 합니다. 인증 서버에 부담이 생길 수 있으며, 클라이언트 수가 많을 경우 세션 관리의 복잡성이 증가할 수 있습니다.
- JWT: JWT는 클라이언트에 정보를 저장하고 서버는 간단한 유효성 검증만 수행하므로, 세션 관리에 대한 부담이 줄어듭니다.


**4. 확장성**
- Session 기반 인증: 서버는 클라이언트의 세션 상태를 유지하기 위해 추가적인 저장소를 사용해야 할 수 있습니다. 따라서 저장소의 확장성이나 관리에 주의가 필요합니다.
-JWT: JWT는 클라이언트에 정보를 저장하므로 추가적인 저장소가 필요하지 않습니다. 따라서 저장소 관리나 확장성에 대한 부담이 적습니다.

Session 기반 인증은 서버 상태를 유지하고 세션 ID를 사용하여 클라이언트를 식별하는 반면, JWT는 자체적으로 상태를 갖지 않고 클라이언트에 정보를 저장하여 인증을 처리합니다.
따라서 JWT는 확장성과 분산 시스템에서 유리하며, 서버의 부담이 적어지는 장점을 가지고 있습니다.
